algorithms/src/snark/varuna/data_structures/proof.rs

```diff
@@ -251,19 +251,45 @@ impl<E: PairingEngine> Proof<E> {
         pc_proof: sonic_pc::BatchLCProof<E>,
     ) -> Result<Self, SNARKError> {
         let batch_sizes: Vec<usize> = batch_sizes.into_values().collect();
-        let total_instances = batch_sizes.iter().sum::<usize>();
-        if commitments.witness_commitments.len() != total_instances {
-            return Err(SNARKError::BatchSizeMismatch);
-        }
         Ok(Self { batch_sizes, commitments, evaluations, third_msg, fourth_msg, pc_proof })
     }
 
     pub fn batch_sizes(&self) -> Result<&[usize], SNARKError> {
+        self.check_batch_sizes()?;
```

Comment:
I think this should instead be called before the current callsite of `batch_sizes`; otherwise we may perform multiple redundant checks in the future (as `batch_sizes` doesn't suggest that any checks are being performed)

Reply:
Yeah was thinking the same. I guess the rationale for calling it from `batch_sizes()` is some sort of assurance that the check will be called... But I moved it into `varuna.rs` and future developers will not be expected to gratitiously remove that check.